packages/roosterjs-editor-dom/lib/userPosition/adjustUserPosition.ts

```diff
@@ -0,0 +1,238 @@
+import {
+    EditorCore,
+    NodePosition,
+    NodeType,
+    PositionType,
+    QueryScope,
+} from 'roosterjs-editor-types';
+import {
+    changeElementTag,
+    contains,
+    createRange,
+    findClosestElementAncestor,
+    getBlockElementAtNode,
+    getTagOfNode,
+    isNodeEmpty,
+    isPositionAtBeginningOf,
+    isVoidHtmlElement,
+    LinkInlineElement,
+    Position,
+    PositionContentSearcher,
+    queryElements,
+    splitBalancedNodeRange,
+    splitTextNode,
+    toArray,
+    unwrap,
+    VTable,
+    wrap,
+} from 'roosterjs-editor-dom';
+
+/**
+ * Adjust position for <a> tag don't be nested inside another <a> tag.
+ */
+
+export function adjustInsertPositionForHyperLink(
+    root: HTMLElement,
+    nodeToInsert: Node,
+    position: NodePosition,
+    core: EditorCore
+): NodePosition {
+    let blockElement = getBlockElementAtNode(root, position.node);
+
+    if (blockElement) {
+        // Find the first <A> tag within current block which covers current selection
+        // If there are more than one nested, let's handle the first one only since that is not a common scenario.
+        let anchor = queryElements(
+            root,
+            'a[href]',
+            null /*forEachCallback*/,
+            QueryScope.OnSelection,
+            createRange(position)
+        ).filter((a: HTMLElement) => blockElement.contains(a))[0];
+
+        // If this is about to insert node to an empty A tag, clear the A tag and reset position
+        if (anchor && isNodeEmpty(anchor)) {
+            position = new Position(anchor, PositionType.Before);
+            safeRemove(anchor);
+            anchor = null;
+        }
+
+        // If this is about to insert nodes which contains A tag into another A tag, need to break current A tag
+        // otherwise we will have nested A tags which is a wrong HTML structure
+        if (
+            anchor &&
+            (<ParentNode>(nodeToInsert as HTMLElement))?.querySelector &&
+            (<ParentNode>(nodeToInsert as HTMLElement))?.querySelector('a[href]')
```

Comment:
You can check safeInstanceOf(nodeToInsert, 'HTMLElement'), then no need to cast again.